keps/sig-apps/3329-retriable-and-non-retriable-failures/README.md

```diff
@@ -822,23 +880,31 @@ condition with dedicated `type`, `reason` and `message` fields based on the
 invocation context.
 
 We list the reason constants above in order to gain the community consensus on
-informative names. It is not supported to use the `reason` or `message` fields
+informative names, but they are purely informational. It is not supported to use the `reason` or `message` fields
 when defining a pod failure policy (see `PodFailurePolicyOnPodConditionsPattern`
 in [JobSpec API](#jobspec-api)). Supporting of matching by the `reason` field
 is a possible extension of the feature which can be implemented once there is
 use case to motivate it. However, it may create a risk of breaking compatibility
 with evolving set of reasons in use, similar to the risk of
 [evolving condition types](#evolving-condition-types).
 
-### Detection of exceeded memory limit
+### Detection and handling of exceeded limits
 
-We modify kubelet to add the `ResourceLimitExceeded` condition based on the
+We modify kubelet to add the `ResourceLimitExceeded` condition if the memory
+or ephemeral-storage limits are exceeded.
+
+In particular, we detect that the memory limits were execeed based on the
 container's `reason` field being equal to `OOMKilled`. This value is set on
```

Comment:
We need more details on when we add this condition.

OOMKilled could happen for other reasons. For example, imagine this scenario:
The Node has 2Gb available, pod A has requests=limits=1Gb, pod B has no requests, but 2Gb limit. If pod A is using the entire 1Gb, and pod B uses more than 1 but less than 2Gb, then it might still get killed, even though it's not exceeding the limit.

Maybe we can say that the condition will only be added if requests and limits are set and they are equal.

Or maybe we can rename the PodCondition to `NodeResourcePressure`? Then it should apply in all cases, and users can decide whether they want to use this condition if they know that they only have pods with guaranteed QoS?

Reply:
Lets change the name to better indicate that this covers all cases, there is no reason to limit this condition to guaranteed QoS pods. I was already assuming that limit in `ResourceLimitExceeded` was meant to indicate "resources the pod is eligible for", but indeed better to use a different name.

`NodeResourcePressure` is not a good name because the pod will be evicted if it exceeds its memory limits even if the node wasn't under memory pressure.

may be `EligibleResourcesExceeded` or `ResourceCapExceeded`?

Reply:
> We need more details on when we add this condition.
> 
> OOMKilled could happen for other reasons. For example, imagine this scenario: The Node has 2Gb available, pod A has requests=limits=1Gb, pod B has no requests, but 2Gb limit. If pod A is using the entire 1Gb, and pod B uses more than 1 but less than 2Gb, then it might still get killed, even though it's not exceeding the limit.

In this scenario there will be a NodeDiskPressure condition added to the node (as >2Gb) is used. Then, the eviction manager tries to delete a Pod to find space. It ranks the Pods according to the function: https://github.com/kubernetes/kubernetes/blob/918f84cd90d121c6624fca5c6eb7b36335030ea0/pkg/kubelet/eviction/helpers.go#L639. Assuming both A & B have the same priority B will be deleted as it uses more disk space. In the KEP (as currently) we will add the DisruptionTarget condition to the Pod with the “DeletedByKubelet” reason which I think is reasonable.

Reply:
> Or maybe we can rename the PodCondition to `NodeResourcePressure`? Then it should apply in all cases, and users can decide whether they want to use this condition if they know that they only have pods with guaranteed QoS?

I don’t like `NodeResourcePressure` for two reasons:
- `Resource` in name suggests it also covers disk, but ephemeral-storage limits are covered by an explicit check in kubelet
- `Pressure` in name suggests it is due to node pressure, but we propose to handle node-pressure eviction by adding DisruptionTarget

If we just want to change the name, then I suggest `OOMKilled`.

Then we would have for memory: 
- OOMKilled (renamed from ResourceLimitExceeded)
- DisruptionTarget for node memory pressure eviction
For ephemeral storage we would have:
- EphemeralStorageLimitExceeded  (renamed from ResourceLimitExceeded), if the explicit check is violated
- DisruptionTarget for node disk pressure eviction

Alternatively, if we want to still have the generic `ResourceLimitExceeded` condition, then I suggest:
- OOMKilled + ResourceLimitExceeded if requests.memory = limits.memory for the killed container  (based on the suggested check for limits = requests)
- OOMKilled only if limits.memory < requests.memory or requests.memory undefined for the killed container
- DisruptionTarget when deleted due to memory pressure
For ephemeral storage without changes:
- ResourceLimitExceeded - if the explicit check is violated
- DisruptionTarget for node disk pressure eviction

Reply:
Let's summarize the current discussion with the list of currently leading approaches:

1. rename `ResourceLimitExceeded` to account for the fact that the limits may not necessarily be exceeded in case of memory as OOM killer can kill a container before it actually exceeds its limit. Proposed names include: `NodeResourcePressure`, `EligibleResourcesExceeded`, `ResourceCapExceeded`. Let me add `ResourcePressure`, `ResourceLimitPressure`.

2. rename  `ResourceLimitExceeded` to `OOMKilled` for memory and to `EphemeralStorageLimitExceeded` for disk. Alternative names which do not leak the OOM-based implementation include: `MemoryPressure`, `MemoryLimitPressure`.

3. continue to use `ResourceLimitExceeded`, but introduce a special case when `resources.limits` is unspecified or greater than `resources.requests` for memory. In that case introduce a dedicated condition reflecting our uncertainty about the limits being actually exceeded. Proposed names: `OOMKilled` (leaks the implementation detail), `MemoryPressure` (preferred), `MemoryLimitPressure`, `ResourceLimitPressure`, `ResourceLimitAboutBeingExceeded`.

My preference goes for 3, because we keep the generic name which is descriptive and accurate for both memory and disk. For the situation when we are uncertain we introduce a dedicated condition. 

@bobbypage @ahg-g @alculquicondor  please share your view / knowledge / opinions.

Reply:
Let's start with what's clear:
When the kubelet (as opposed to the kernel) evicts the pod before it reaches the limit (for memory and storage), we should add a `DisruptionTarget` condition. Probably with a reason along the lines of `NodeMemoryPressure`, `NodeEphemeralStoragePressure`.

In general, I would prefer that we keep a small set of conditions to make it easier for users to write Job pod failure policies. I think option 1 is closer to the ideal. However, it's indeed hard to find a name that doesn't use the word pressure (that would conflict with the reasons above) and encompasses both storage and memory. Maybe the solution is something a bit less intuitive, like: `ResourceReclaimed` or `ResourceExhausted` (note the absence of Limit in the name), with reasons `OOMKilled`, `EphemeralStoragePressure` (this case would also get a `DisruptionTarget` condition) and `EphemeralStorageLimit`. Then we can document that when requests.memory=limits.memory, it is guaranteed that this condition is equivalent to "limit exceeded".

If we can't agree on the names, then the way to go would be to have more specific conditions, in which case option 2 is probably better.

Reply:
I discussed with @SergeyKanzhelev, and he mentioned that the memory pressure logic we have in kubelet is based on a metric scraped from cadvisor, the granularity of scraping is ~1min, so there is a good chance that the kernel will be the one acting on killing the pod when sudden spikes happen.

With that being said, I would focus on what the user will be trying to do to guide us in how we build the API: In general, users will likely want to fail jobs when their pods get killed because of exceeding the memory limit they set (because likely this indicates a bug on their side), but retry when they get killed because of memory pressure.

For pods killed by the kernel we can't distinguish between the case where they were killed because of limit exceeded or because of memory pressure; therefore, whether or not we have a separate condition for OOMKilled will not really help the user because they will likely set a policy to fail the job on OOMKill condition anyway plus the EphemeralStorageLimitExceeded. So might as well just have a single condition for both (option 1).

In summary, I think a single condition (option 1) will be simpler than being more granular (option 2), yet just as useful.

Reply:
Ok, let's explore option 1. What would be the name? Out of the proposed names my preference goes to `ResourceExhausted`.

Reply:
maybe `NodeResourceExhausted` to be more specific.

Reply:
It will be the pod's resources that were exhausted when it exceeds its limits, not the node. I would go with `ResourceExhausted` alone.

Reply:
I prefer `ResourceExhaused` too, renamed.